### ISSUE TITLE
Transformer.generate_square_subsequent_mask has nan values on MPS device

### ISSUE THREAD

**Issue by psandovalsegura** [pytorch/pytorch#106341]
### 🐛 Describe the bug

Square subsequent mask has nan values in place of zeroes when created on MPS device. To reproduce:

```
import torch.nn as nn
print(nn.modules.transformer.Transformer.generate_square_subsequent_mask(4, device='mps'))
```
prints:
```
tensor([[nan, -inf, -inf, -inf],
        [nan, nan, -inf, -inf],
        [nan, nan, nan, -inf],
        [nan, nan, nan, nan]], device='mps:0')
```

However, the correct tensor is printed when generated on CPU,  `nn.modules.transformer.Transformer.generate_square_subsequent_mask(4, device='cpu')` prints:
```
tensor([[0., -inf, -inf, -inf],
        [0., 0., -inf, -inf],
        [0., 0., 0., -inf],
        [0., 0., 0., 0.]])
```

After looking into the `Transformer.generate_square_subsequent_mask` function, the problem appears to come from `torch.triu` when called with `diagonal=1`, since `torch.full` works as expected on MPS device.

### Versions

PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.4 (arm64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.17 (main, Jul  5 2023, 15:35:09)  [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Max

Versions of relevant libraries:
[pip3] numpy==1.25.0
[pip3] torch==2.0.1
[pip3] torchvision==0.15.2
[conda] numpy                     1.25.0                   pypi_0    pypi
[conda] torch                     2.0.1                    pypi_0    pypi
[conda] torchvision               0.15.2                   pypi_0    pypi
```[tasklist]
### Tasks
```

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by jeatonsf** [comment]
Also just ran into this very specific bug
```
>>> torch.triu(torch.full((3, 3), float("-inf"), device="cpu"), diagonal=1)
tensor([[0., -inf, -inf],
        [0., 0., -inf],
        [0., 0., 0.]])

>>> torch.triu(torch.full((3, 3), float("-inf"), device="mps"), diagonal=1)
tensor([[nan, -inf, -inf],
        [nan, nan, -inf],
        [nan, nan, nan]], device='mps:0')
```

**Comment by jhavukainen** [comment]
This triu issue was fixed in 2.5.0 by https://github.com/pytorch/pytorch/pull/128575
